sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala

```diff
@@ -125,15 +125,15 @@ case class PlanSubqueries(sparkSession: SparkSession) extends Rule[SparkPlan] {
 case class ReuseSubquery(conf: SQLConf) extends Rule[SparkPlan] {
 
   def apply(plan: SparkPlan): SparkPlan = {
-    if (!conf.exchangeReuseEnabled) {
+    if (!conf.subqueryReuseEnabled) {
       return plan
     }
     // Build a hash map using schema of subqueries to avoid O(N*N) sameResult calls.
     val subqueries = mutable.HashMap[StructType, ArrayBuffer[SubqueryExec]]()
     plan transformAllExpressions {
       case sub: ExecSubqueryExpression =>
         val sameSchema = subqueries.getOrElseUpdate(sub.plan.schema, ArrayBuffer[SubqueryExec]())
-        val sameResult = sameSchema.find(_.sameResult(sub.plan))
+        val sameResult = sameSchema.find(_.child.sameResult(sub.plan.child))
```

Comment:
This change fixes the bug you described in the PR description? Sorry, but I'm not clearly sure about what's the problem here, so could you show us an example query to reproduce the bug?

Reply:
Thanks for the review, you can see the example in the added test.

Reply:
I am not sure either about this fix. I remember this problem which I tried to fix in https://github.com/apache/spark/pull/22518/commits/36fa664c6d251901270984115ff2ebfd1b665fca. As you can see, the PR didn't actually do this fix as per @cloud-fan 's comment. So I am pinging him here for reviewing this too. Anyway, when I debugged it, the problem was that being the same object according to equals, the tree was not updated, so I see no reason for the current fix.

Reply:
shall we just overwrite `SubqueryExec.doCanonicalize` to strip it? It seems the `SubqueryExec.name` is stopping us from match semantically same subqueries.

Reply:
I have updated code as @cloud-fan suggests, thanks for all your reviews!